src/backend/optimizer/plan/createplan.c

```diff
@@ -6471,6 +6475,132 @@ make_modifytable(PlannerInfo *root,
 	return node;
 }
 
+/*
+ * can_elide_explicit_motion_recurse
+ *	  Recursive guts of can_elide_explicit_motion().
+ *
+ *	  Check for motions by recursively looking at right subtree first. If it does
+ *	  not exist, go into left subtree. If we find a motion before the scan, we
+ *	  return false and exit from that subtree. A scan is always underneath the
+ *	  motion, so we won't encounter the scan if we exited the subtree. In that
+ *	  case, we can't elide Explicit Redistribute Motion.
+ *
+ *	  If we encounter the scan before any motions, then we can elide
+ *	  unneccessary Explicit Redistribute Motion.
+ */
+static bool
+can_elide_explicit_motion_recurse(Plan *plan, List *rtable, Oid relid,
+								  bool relid_is_subclass, Oid reltypeid)
+{
+	while (plan)
+	{
+		/*
+		 * If this is a Motion, check Oids from the Motion's targetlist.
+		 */
+		if (IsA(plan, Motion))
+		{
+			Motion *motion = (Motion *) plan;
+
+			if ((motion->motionType == MOTIONTYPE_FIXED && motion->isBroadcast) ||
+				motion->motionType == MOTIONTYPE_HASH)
+			{
+				ListCell *lcr;
+
+				foreach(lcr, plan->targetlist)
+				{
+					TargetEntry *tle = (TargetEntry *) lfirst(lcr);
+					RangeTblEntry *target_rte = rt_fetch(tle->resno, rtable);
+
+					if (relid_is_subclass)
+					{
+						Oid target_reltypeid = get_rel_type_id(target_rte->relid);
+
+						/*
+						 * Does relid inherit from a table in targetlist?
+						 */
+						if (target_reltypeid != InvalidOid && reltypeid != InvalidOid &&
+							typeInheritsFrom(reltypeid, target_reltypeid))
+							/* There is a Motion on parent table before scan on
+							   the child */
+							return false;
+					}
+
+					if (target_rte->relid == relid)
+						/* There is a Motion before scan */
+						return false;
+				}
+			}
+		}
+		/*
+		 * If this is a scan and it's scanrelid matches relid, we encountered a
+		 * scan before any Motions.
+		 */
+		else
+		{
+			switch (nodeTag(plan))
+			{
+				case T_SeqScan:
+				case T_DynamicSeqScan:
+				case T_ExternalScan:
+				case T_IndexScan:
+				case T_DynamicIndexScan:
+				case T_IndexOnlyScan:
+				case T_BitmapIndexScan:
+				case T_DynamicBitmapIndexScan:
+				case T_BitmapHeapScan:
+				case T_DynamicBitmapHeapScan: {
+					Scan *scan = (Scan *) plan;
+					RangeTblEntry *target_rte = rt_fetch(scan->scanrelid, rtable);
+
+					if (target_rte->relid == relid)
+						/* There wasn't any Motions before scan */
+						return true;
```

Comment:
I think it's better to use explicit braces, or move the comment before the condition.